File: resources/xcode/NativePHP/NativeRender/NativeNodeDecoratorRegistry.swift
import SwiftUI

/// Ordered, opt-in view decorators supplied by native plugins.
final class NativeNodeDecoratorRegistry {
    static let shared = NativeNodeDecoratorRegistry()
    typealias Decorator = (_ node: NativeUINode, _ content: AnyView) -> AnyView

    private var decorators: [String: Decorator] = [:]
    private var order: [String] = []
    private var pipeline: Decorator?
    private let lock = NSLock()

    var currentPipeline: Decorator? {
        lock.lock(); defer { lock.unlock() }
        return pipeline
    }

    private init() {}

    func register(_ name: String, decorator: @escaping Decorator) {
        lock.lock(); defer { lock.unlock() }
        if decorators[name] == nil { order.append(name) }
        decorators[name] = decorator
        rebuildPipelineLocked()
    }

    func unregister(_ name: String) {
        lock.lock(); defer { lock.unlock() }
        decorators.removeValue(forKey: name)
        order.removeAll { $0 == name }
        rebuildPipelineLocked()
    }

    private func rebuildPipelineLocked() {
        let snapshot = order.compactMap { decorators[$0] }
        guard !snapshot.isEmpty else {
            pipeline = nil
            return
        }

        pipeline = { node, content in
            snapshot.reduce(content) { view, decorator in decorator(node, view) }
        }
    }
}

File: resources/androidstudio/app/src/main/java/com/nativephp/mobile/ui/nativerender/NativeNodeDecoratorRegistry.kt
package com.nativephp.mobile.ui.nativerender

import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.write

typealias NativeNodeDecorator = @Composable (NativeUINode, Modifier) -> Modifier

/** Ordered, opt-in modifier decorators supplied by native plugins. */
object NativeNodeDecoratorRegistry {
    private val lock = ReentrantReadWriteLock()
    private val decorators = linkedMapOf<String, NativeNodeDecorator>()
    @Volatile private var snapshot: List<NativeNodeDecorator> = emptyList()

    fun register(name: String, decorator: NativeNodeDecorator) {
        lock.write {
            decorators[name] = decorator
            snapshot = decorators.values.toList()
        }
    }

    fun unregister(name: String) {
        lock.write {
            decorators.remove(name)
            snapshot = decorators.values.toList()
        }
    }

    fun isEmpty(): Boolean = snapshot.isEmpty()

    @Composable
    fun apply(node: NativeUINode, modifier: Modifier): Modifier {
        var decorated = modifier
        for (decorator in snapshot) decorated = decorator(node, decorated)
        return decorated
    }
}
